### PR TITLE
Use `ObjectMeta.GenerateName` while creating HPA and VPA resources

### LINKED ISSUE #22: [BUG] Use `ObjectMeta.GenerateName` while creating HPA and VPA resources
**Describe the bug:**
At present, the HVPA controller creates HPA and VPA resources based on the templates in the `HVPASpec` by adding the hard-coded [`-hpa`](https://github.com/gardener/hvpa-controller/blob/master/pkg/controller/hvpa/hvpa_controller.go#L590) and [`-vpa`](https://github.com/gardener/hvpa-controller/blob/master/pkg/controller/hvpa/hvpa_controller.go#L538) suffixes to the HVPA name correspondingly.

**Expected behaviour:**
Now with the full template in the `HVPASpec` with including the selector and the metadata for both [HPA](https://github.com/gardener/hvpa-controller/pull/19/files#diff-617e45ba1eda37bfddd8607785d008c6R152) and [VPA](https://github.com/gardener/hvpa-controller/pull/19/files#diff-617e45ba1eda37bfddd8607785d008c6R110), we should use the `ObjectMeta.GenerateName` functionality of `apimachinery` to avoid the hard-coded suffixes.

We should also use the template metadata and the selectors according to the best-practices for embedded templates for generated aggregated resources.

**How To Reproduce (as minimally and precisely as possible):**

**Logs:**

**Screenshots (if applicable):**

**Environment (please complete the following information):**
 - #19 

**Anything else we need to know?**:
We should fix this before we integrate with gardener.

### PR DESCRIPTION
**Pull request by ggaurav10**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #22 

**Special notes for your reviewer**:

**Release note**:

```improvement operator

```

### CODE REVIEW

File: controllers/hvpa_controller.go [hunk 1]
```diff
@@ -280,47 +274,67 @@ func (r *HvpaReconciler) deleteHvpaFinalizers(hvpa *autoscalingv1alpha1.Hvpa) {
...
+		return nil, err
 	}
 
-	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("vpa template in hvpa object already has an owner reference")
+	// list all vpas to include the vpas that don't match the hvpa`s selector
+	// anymore but has the stale controller ref.
+	vpas := &vpa_api.VerticalPodAutoscalerList{}
+	err = r.List(context.TODO(), vpas, client.InNamespace(hvpa.Namespace))
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]:
```suggestion
	listOpts := &client.ListOptions{
	  LabelSelector: selector,
	  Namespace: hvpa.Namespace,
	}
	err = r.List(context.TODO(), vpas, client.InNamespace(hvpa.Namespace), client.UseListOptions(listOpts))
```

With this change selector is not required for filtering later.

File: controllers/hvpa_controller.go [hunk 2]
```diff
@@ -280,47 +274,67 @@ func (r *HvpaReconciler) deleteHvpaFinalizers(hvpa *autoscalingv1alpha1.Hvpa) {
...
 	}
 
-	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
-		log.V(3).Info("Warning", "Generate name provided in the vpa template will be ignored", generateName)
+	updatePolicy := hvpa.Spec.Vpa.UpdatePolicy.UpdateMode
+
+	// NOTE: filteredVpas are pointing to deepcopies of the cache, but this could change in the future.
+	// Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.2/pkg/cache/internal/cache_reader.go#L74
+	// if you need to modify them, you need to copy it first.
+	filteredVpas, err := r.claimVpas(hvpa, selector, vpas)
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Can we filtering here by filtering during the `List` call [above](https://github.com/gardener/hvpa-controller/pull/34/files#diff-6354ac486a904d10093c497f09f83a3eR287)?

File: controllers/hvpa_controller.go [hunk 3]
```diff
@@ -330,120 +344,85 @@ func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.
...
+		return nil, err
 	}
 
-	if ownerRef := metadata.GetOwnerReferences(); len(ownerRef) != 0 {
-		// TODO: Could be done better as part of validation
-		return nil, fmt.Errorf("hpa template in hvpa object already has an owner reference")
+	// list all hpas to include the hpas that don't match the hvpa`s selector
+	// anymore but has the stale controller ref.
+	hpas := &autoscaling.HorizontalPodAutoscalerList{}
+	err = r.List(context.TODO(), hpas, client.InNamespace(hvpa.Namespace))
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]:
```suggestion
	listOpts := &client.ListOptions{
	  LabelSelector: selector,
	  Namespace: hvpa.Namespace,
	}
	err = r.List(context.TODO(), hpas, client.InNamespace(hvpa.Namespace), client.UseListOptions(listOpts))
```

With this change selector is not required for filtering later.

File: controllers/hvpa_controller.go [hunk 4]
```diff
@@ -330,120 +344,85 @@ func (r *HvpaReconciler) reconcileVpa(hvpa *autoscalingv1alpha1.Hvpa) (*vpa_api.
...
 	}
 
-	if generateName := metadata.GetGenerateName(); len(generateName) != 0 {
-		log.V(3).Info("Warning", "Generate name provided in the hpa template will be ignored", generateName)
+	updatePolicy := hvpa.Spec.Hpa.UpdatePolicy.UpdateMode
+
+	// NOTE: filteredHpas are pointing to deepcopies of the cache, but this could change in the future.
+	// Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.2/pkg/cache/internal/cache_reader.go#L74
+	// if you need to modify them, you need to copy it first.
+	filteredHpas, err := r.claimHpas(hvpa, selector, hpas)
```
Comment (amshuman-kr) [inline controllers/hvpa_controller.go]: Can we avoid this filtering by filtering in the `List` call [above](https://github.com/gardener/hvpa-controller/pull/34/files#diff-6354ac486a904d10093c497f09f83a3eR364)?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Can we filter in the `List` call and avoid separate filtering?

**Comment by ggaurav10** [comment]
The `List` call lists all HPAs/VPAs in the namespace, and then does the filtering later, because there can be cases where `ownerRef` is present, but `labels` don't match anymore. In these cases, we would want to [`release`](https://github.com/gardener/hvpa-controller/blob/5e7be193561126ad5a4d7ffe21bf12fd0960001b/controllers/controller_ref_manager.go#L82-L110) those HPAs/VPAs.
Will "resolve" the related conversations. Please "unresolve" if needed.

**Comment by amshuman-kr** [comment]
> The List call lists all HPAs/VPAs in the namespace, and then does the filtering later, because there can be cases where ownerRef is present, but labels don't match anymore. In these cases, we would want to release those HPAs/VPAs.
Will "resolve" the related conversations. Please "unresolve" if needed.

👍 Forgot about that. LGTM provided other comments are addressed.

**Review by amshuman-kr — APPROVED** [review]
LGTM

**Pull request merged**